app/controllers/parent_objects_controller.rb

```diff
@@ -59,8 +60,22 @@ def valid_request?
 
   # PATCH/PUT /parent_objects/1
   # PATCH/PUT /parent_objects/1.json
+  # rubocop:disable Metrics/AbcSize
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/MethodLength
+  # rubocop:disable Metrics/PerceivedComplexity
   def update
     respond_to do |format|
+      parent_object = ParentObject.find(params[:id])
+      permission_set = parent_object&.permission_set
+      permission_set_param = OpenWithPermission::PermissionSet.find(parent_object_params[:permission_set_id]) if parent_object_params[:permission_set_id].present?
+
+      authorize!(:owp_access, permission_set) if parent_object.visibility == "Open with Permission" && parent_object.visibility != parent_object_params[:visibility]
+
+      authorize!(:owp_access, permission_set) if permission_set.present? && permission_set_param.nil?
```

Comment:
I think this check should be performed whenever the permission_set is present and is not equal to the permission_set_param (i.e., whether `nil` or the assignment is to a different permission set).